url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/TextField.java

```diff
@@ -0,0 +1,347 @@
+
+	public TextField(float x, float y, float w) {
+		super(x, y, w, 0);
+	}
+	
+	public TextField(float x, float y, float w, String initialText) {
+		this(x, y, w);
+		this.text = initialText;
+	}
+	
+	public String getText() {
+		return text;
+	}
+
+	public void setText(String text) {
+		TextEvent e = new TextEvent(this);
+		e.setOldText(getText());
+		e.setNewText(text);
+		
+		this.text = text;
+		
+		fireTextChanged(e);
+	}
+
+	public int getCaretIndex() {
+		return caretIndex;
+	}
+
+	public void setCaretIndex(int caretIndex) {
+		if (caretIndex < 0) caretIndex = 0;
+		if (caretIndex > getText().length()) caretIndex = getText().length();
+		this.caretIndex = caretIndex;
+	}
+
+	public int getSecondaryCaretIndex() {
+		return secondaryCaretIndex;
+	}
+
+	public void setSecondaryCaretIndex(int secondaryCaretIndex) {
+		if (secondaryCaretIndex < 0) secondaryCaretIndex = 0;
+		if (secondaryCaretIndex > getText().length()) secondaryCaretIndex = getText().length();
+		this.secondaryCaretIndex = secondaryCaretIndex;
+	}
+
+	@Override
+	public Color getFillColor() {
+		return TEXT_FIELD_FILL_COLOR;
+	}
+	
+	@Override
+	public Color getStrokeColor() {
+		return TEXT_FIELD_BORDER_COLOR;
+	}
+	
+	@Override
+	public float getStrokeWeight() {
+		return TEXT_FIELD_BORDER_WIDTH;
+	}
+	
+	@Override
+	public void preRender(PApplet p) {
+		super.preRender(p);
+		setHeight(p.textAscent()+p.textDescent()+2*TEXT_FIELD_TEXT_PADDING);
+	}
+	
+	@Override
+	public void render(PApplet p) {
+		super.render(p);
+		
+		if (textCanvas == null) {
+			textCanvas = p.createGraphics((int) getWidth()-2*TEXT_FIELD_TEXT_PADDING, (int) getHeight()-2*TEXT_FIELD_TEXT_PADDING, PApplet.JAVA2D);
+		}
+		textCanvas.beginDraw();
+		textCanvas.stroke(TEXT_FIELD_TEXT_COLOR.getRGB());
+		textCanvas.fill(TEXT_FIELD_TEXT_COLOR.getRGB());
+		textCanvas.textAlign(PApplet.LEFT, PApplet.TOP);
+		textCanvas.textFont(NUIController.getInstance().getFont());
+		textCanvas.rectMode(PApplet.CORNER);
+		
+		textCanvas.background(TEXT_FIELD_FILL_COLOR.getRGB());
+		float caretPos = textCanvas.textWidth(getText().substring(0, getCaretIndex()));
+		if (caretPos + textOffset > textCanvas.width) {
+			textOffset -= caretPos + textOffset - textCanvas.width + 2;
+		}
+		if (caretPos + textOffset < 0) {
+			textOffset += -(caretPos + textOffset) + 2;
+		}
+		textCanvas.text(getText(), textOffset, 0);
+		textCanvas.noStroke();
+		textCanvas.fill(16, 16, 255, 80);
+		textCanvas.rectMode(PApplet.CORNERS);
+		float secondaryPos = textCanvas.textWidth(getText().substring(0, getSecondaryCaretIndex()));
+		textCanvas.rect(secondaryPos + textOffset, 0, caretPos + textOffset, textCanvas.height);
+		if (!keyboards.isEmpty() && ++timer%30 < 15) {
+			textCanvas.stroke(TEXT_FIELD_TEXT_COLOR.getRGB());
+			textCanvas.line(caretPos + textOffset, 0, caretPos + textOffset, textCanvas.height);
+		}
+		textCanvas.endDraw();
+		
+		p.image(textCanvas, 0, 0);
+	}
+
+	public void keyPressed(KeyEvent e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.